New Controllers, IK Refactoring

cremebrule opened the pull request:
Adding support for following controllers:
EE Impedance (Position, Position & Orientation)
Joint Impedance
Joint Velocity
Joint Torque

Also refactoring of IK controller to fit new controller architecture

amandlek commented on robosuite/controllers/base_controller.py:
```diff
@@ -0,0 +1,98 @@
+    ):
+        self.action_scale = None
+
+        # Note: the following attributes must be specified in extended controller subclasses during the init call:
```
It might be useful to make these class properties (method with the @property tag) and make them not implemented in this base class, so that subclasses are forced to implement these.

amandlek replied:
It might be useful to make these class properties (method with the @ property tag) and make them not implemented in this base class, so that subclasses are forced to implement these.

amandlek commented on robosuite/controllers/base_controller.py:
```diff
@@ -0,0 +1,98 @@
+
+        self.model_timestep = self.sim.model.opt.timestep
+
+        self.ee_pos = self.sim.data.body_xpos[self.sim.model.body_name2id(self.id_name)]
```
for all of these where you are reading from self.sim, it might be safer to wrap these in a np.array call to ensure copies are made

amandlek replied:
for all of these where you are reading from self.sim, it might be safer to wrap these in a np.array call to ensure copies are made

amandlek commented on robosuite/controllers/config/ee_pos.json:
```diff
@@ -0,0 +1,16 @@
+  "kp_min": 10,
+  "impedance_flag": false,
+  "use_delta_impedance": false,
+  "interpolation": "linear"
```
is there a reason that the interpolation is linear here but null in other configs?

amandlek commented on robosuite/controllers/config/joint_imp.json:
```diff
@@ -0,0 +1,12 @@
+  "damping_min": [0, 0, 0, 0, 0, 0, 0],
+  "impedance_flag": false,
+  "use_delta_impedance": false,
+  "interpolation": "linear"
```
same here, why linear interpolation?

amandlek commented on robosuite/controllers/config/joint_torque.json:
```diff
@@ -0,0 +1,7 @@
+  "control_range": [0.5, 0.5, 0.5, 0.2, 0.2, 0.1, 0.1],
+  "impedance_flag": false,
+  "use_delta_impedance": false,
+  "interpolation": "linear"
```
same here, why linear interpolation?

amandlek commented on robosuite/controllers/ee_imp.py:
```diff
@@ -0,0 +1,164 @@
+            self.relative_ori = np.zeros(3) #relative orientation always starts at 0
+
+    def run_controller(self):
+        # TODO: check if goal has been set.
```
what is this TODO here?

amandlek commented on robosuite/environments/base.py:
```diff
@@ -195,15 +195,19 @@ def step(self, action):
-        self._pre_action(action)
-        end_time = self.cur_time + self.control_timestep
-        while self.cur_time < end_time:
+        policy_step = True
```
comments for what the `policy_step` variable does would be useful - explain that the env.step frequency is different from the number of simulation steps since a controller will output torques multiple times for a single policy output

amandlek replied:
@cremebrule Please address this comment

cremebrule replied:
Sorry -- missed this comment. done!

amandlek commented on robosuite/environments/sawyer.py:
```diff
@@ -192,39 +219,46 @@ def _pre_action(self, action):
+        # Now run the controller for a step
+        torques = self.controller.run_controller()
+
+        # TODO? No torque compensation currently in controller? (set to zeros it seems)
```
what is this TODO?

amandlek commented on robosuite/environments/sawyer.py:
```diff
@@ -281,8 +315,10 @@ def action_spec(self):
-        low = np.ones(self.dof) * -1.
-        high = np.ones(self.dof) * 1.
+        # Torque limit values pulled from relevant robot.xml file
+        low = np.array([-80., -80., -40., -40., -9., -9., -9.])
```
instead of hardcoding, can we actually read them from the file?

amandlek commented on robosuite/environments/sawyer_lift.py:
```diff
@@ -96,6 +101,19 @@ def __init__(
+            try:
+                with open(controller_path) as f:
+                    controller_config = json.load(f)
+                    # Override the velocity kv parameters to be suited for sawyer
```
can you explain this?

amandlek commented on robosuite/environments/sawyer_lift.py:
```diff
@@ -189,14 +208,19 @@ def _reset_internal(self):
         Resets simulation internal configurations.
         """
+        # reset joint positions (gets reset in sim during super() call)
+        # TODO: This manually overrides init_qpos found in robot_model, so maybe best to merge these two into one attribute
```
what's going on here?

amandlek commented on robosuite/models/robots/sawyer_robot.py:
```diff
@@ -26,4 +26,6 @@ def joints(self):
 
     @property
     def init_qpos(self):
+        # TODO: This is the attribute that should be writeable
```
what is this TODO, and why is there another set of joints commented out?

amandlek commented on robosuite/utils/transform_utils.py:
```diff
@@ -528,6 +552,45 @@ def rotation_matrix(angle, direction, point=None):
+    if quat[3] == 1:
+        # This is a zero degree rotation, immediately return
+        return quat
+    elif quat[3] == 0:
```
why is this if statement required? it seems like the case below would return the right values. is this due to rounding issues?

amandlek replied:
why is this if statement required? it seems like the case below would return the right values. is this due to rounding issues?

amandlek commented on robosuite/controllers/base_controller.py:
```diff
@@ -0,0 +1,155 @@
+        self.ee_pos_vel = np.array(self.sim.data.body_xvelp[self.sim.model.body_name2id(self.id_name)])
+        self.ee_ori_vel = np.array(self.sim.data.body_xvelr[self.sim.model.body_name2id(self.id_name)])
+
+        self.joint_pos = np.array(self.sim.data.qpos[self.joint_index])
```
@cremebrule This might be using the wrong set of indices - you are using `env._ref_joint_vel_indexes` but we might need `env._ref_joint_pos_indexes` to index into the qpos array. Can you please check?

amandlek commented on robosuite/controllers/README.md:
```diff
@@ -0,0 +1,43 @@
+
+# Import the file as a dict
+with open(controller_fpath) as f:
+	config = json.load(f)
```
can you fix this spacing to be 4 spaces?

cremebrule replied:
done

amandlek commented on robosuite/controllers/ee_imp.py:
```diff
@@ -0,0 +1,180 @@
+                                                                                 self.J_full,
+                                                                                 self.J_pos,
+                                                                                 self.J_ori)
+        if self.uncoupling:
```
could use a comment on what uncoupling does. In general, for this controller implementation, it would be great to reference or add equations in comments for what's going on in the OSC computation (for example, tau = J.T * F, F = -kp(x - x_des) -kv(xdot), etc.) - otherwise people will be lost when reading this code and trying to understand what happens

cremebrule replied:
done

amandlek commented on robosuite/environments/panda.py:
```diff
@@ -91,12 +97,38 @@ def __init__(
+        #   policy (control) freq, and ndim (# joints)
+        controller_config["robot_name"] = self.mujoco_robot.name
+        controller_config["sim"] = self.sim
+        controller_config["robot_id"] = "right_hand"
```
is this supposed to correspond to the name of the robot end effector? If so, can we rename this to something else? I feel like `robot_id` is a little confusing and vague, maybe something like `robot_eef_id`?

cremebrule replied:
I agree -- switched it to "eef_name" to be more precise

amandlek commented on robosuite/models/grippers/two_finger_gripper.py:
```diff
@@ -23,7 +23,7 @@ def init_qpos(self):
     @property
     def joints(self):
-        return ["r_gripper_l_finger_joint", "r_gripper_r_finger_joint"]
+        return ["r_gripper_r_finger_joint", "r_gripper_l_finger_joint"]
```
do you remember why this ordering was reversed? just curious

cremebrule replied:
for some reason the original mapping was contrary to what the gripper docstring dictated. This fix makes it so that the open / close convention as defined by the gripper docstring is consistent across all grippers (action: 1 => open, -1 => closed)

amandlek commented on robosuite/utils/control_utils.py:
```diff
@@ -0,0 +1,167 @@
+import scipy.linalg
+
+
+def nullspace_torques(mass_matrix, nullspace_matrix, initial_joint, joint_pos, joint_vel):
```
please add docstring and comments on what this function does (talk about the nullspace controller that tries to maintain a secondary objective of maintaining a desired joint position)

amandlek replied:
@cremebrule I wanted to follow up on this - do you need any help here?

cremebrule replied:
sorry -- missed all of these comments that were compressed when scrolling through them! done. Please feel free to double-check the docstrings I added and add anything you feel is missing / unclear

amandlek commented on robosuite/utils/control_utils.py:
```diff
@@ -0,0 +1,167 @@
+
+
+def nullspace_torques(mass_matrix, nullspace_matrix, initial_joint, joint_pos, joint_vel):
+    joint_kp = 10
```
should this be an argument with a default value?

amandlek replied:
@cremebrule following up on this

cremebrule replied:
yes! good call

amandlek commented on robosuite/utils/control_utils.py:
```diff
@@ -0,0 +1,167 @@
+
+def nullspace_torques(mass_matrix, nullspace_matrix, initial_joint, joint_pos, joint_vel):
+    joint_kp = 10
+    joint_kv = np.sqrt(joint_kp) * 2
```
comment that this corresponds to critical damping

amandlek replied:
@cremebrule following up on this

cremebrule replied:
done! added line by line comments for this function as well

amandlek commented on robosuite/utils/control_utils.py:
```diff
@@ -0,0 +1,167 @@
+        J_ori.transpose())
+
+    # take the inverse, but zero out elements in cases of a singularity
+    # todo: maybe there is an error here? check with danfei and ajay
```
this todo?

amandlek replied:
@cremebrule following up on this

cremebrule replied:
I'll follow up with michelle about this

amandlek commented on robosuite/utils/control_utils.py:
```diff
@@ -0,0 +1,167 @@
+def set_goal_position(delta,
+                       current_position,
+                       position_limit=None,
+                       set_pos=None):
```
add docstring - what does this function do?

cremebrule replied:
done

amandlek commented on robosuite/utils/control_utils.py:
```diff
@@ -0,0 +1,167 @@
+def set_goal_orientation(delta,
+                         current_orientation,
+                         orientation_limit=None,
+                         set_ori=None):
```
add docstring - what does this function do?

cremebrule replied:
done

amandlek commented on robosuite/utils/control_utils.py:
```diff
@@ -0,0 +1,167 @@
+        if orientation_limit.shape != (2,3):
+            raise ValueError("Orientation limit should be shaped (2,3) "
+                             "but is instead: {}".format(orientation_limit.shape))
+        # TODO: Limit rotation!
```
lingering TODO?

cremebrule replied:
original todo when I ported it over from michelle's repo. Seems like she already had implemented some form of limiting so I'm unsure as to exactly what the todo is referring to. Will follow up with her

cremebrule replied:
resolved

amandlek commented on robosuite/utils/control_utils.py:
```diff
@@ -0,0 +1,167 @@
+
+        limited = False
+        for idx in range(3):
+            if orientation_limit[0][idx] < orientation_limit[1][idx]:  # Normal angle sector meaning
```
what does this whole snippet of code do? it's confusing

cremebrule replied:
I think it effectively checks to make sure the requested goal orientation is within the `orientation_limit` bounds, and then does some bounds checking based on how the limits were specified / the computed orientation goal. Will double check with michelle

cremebrule replied:
followed up with michelle and turns out @roberto-martinmartin is the culprit of this code!! :p

amandlek replied:
@roberto-martinmartin could you please help add some comments here to resolve this? and maybe even break this clipping out into a separate utility function for clipping euler angles?

amandlek commented on tests/test_controllers/test_all_controllers.py:
```diff
@@ -0,0 +1,93 @@
+"""
+Test all controllers on the Sawyer(Lift) task environment as a test case.
+
+The following controllers are tested:
```
What are the test cases? Additional descriptions would be helpful!

cremebrule replied:
Added significant documentation!

amandlek replied:
@cremebrule minor comment about some documentation - you say "e.g. the pos/ori controller should be expected to move in the x direction first, then the y direction, ..." - this description is a little confusing, what do you mean by that? it suggests that the controller is moving axis by axis sequentially

amandlek replied:
Is this what the test is supposed to do to the arm?

cremebrule replied:
yes! exactly. I've added more documentation that is more clear and explicit

amandlek commented on tests/test_controllers/test_all_controllers.py:
```diff
@@ -0,0 +1,93 @@
+            count += 1
+
+        # Tests passed!
+        print("All controller tests passed successfully!")
```
do you check for any expected output? is there some qualitative end condition that a user of this script should check for?

cremebrule replied:
This is exclusively a qualitative test -- descriptions of each controller's expected behavior as well as a description of the general scope of the testing module has been written in the docstring at the top

yukezhu commented on robosuite/controllers/README.md:
```diff
@@ -0,0 +1,43 @@
+Below, a brief overview and description of each subset of controller parameters are shown:
+
+#### General Controller Settings  
+* `'type'`: Type of controller to control. Can be `'EE_POS_ORI'`, `'EE_POS'`, `'EE_IK'`, `'JOINT_IMP'`, `'JOINT_VEL'`, or `'JOINT_TORQUE'`
```
What controller settings are mandatory and what are optional? And what are their default values?

cremebrule replied:
As mentioned in the Configuration section on line 13, the default templates provided in the `config` folder describe both the default values and the required parameters that should be specified.

There are additional default config files for each robot arm -- added a blurb that mentions that these specific default jsons are used by default for each robot, respectively

yukezhu commented on robosuite/controllers/base_controller.py:
```diff
@@ -0,0 +1,157 @@
+        return transformed_action
+
+    def update(self):
+
```
Add documentation for the class methods below

cremebrule replied:
done

yukezhu commented on robosuite/controllers/ee_imp.py:
```diff
@@ -0,0 +1,199 @@
+
+class EEImpController(Controller):
+    """
+    Controller for EE impedance
```
More high-level description of the EE impedance controller

yukezhu replied:
Explain the input arguments (would be good to include some equations how they are used) -- and similar for the other controllers

cremebrule replied:
done. equation notes (as well as reference to the original [Khatib paper](http://khatib.stanford.edu/publications/pdfs/Khatib_1987_RA.pdf)) have been included in the `run_controller` method

yukezhu commented on robosuite/controllers/joint_imp.py:
```diff
@@ -0,0 +1,114 @@
+
+class JointImpController(Controller):
+    """
+    Controller for joint impedance
```
Same as EE impedance controller, add more documentation

cremebrule replied:
done

yukezhu commented on robosuite/controllers/joint_torque.py:
```diff
@@ -0,0 +1,99 @@
+
+class JointTorqueController(Controller):
+    """
+    Controller for joint torque
```
Same as EE impedance controller, add more documentation

cremebrule replied:
done

yukezhu commented on robosuite/controllers/joint_vel.py:
```diff
@@ -0,0 +1,105 @@
+
+class JointVelController(Controller):
+    """
+    Controller for joint velocity
```
Same as EE impedance controller, add more documentation

cremebrule replied:
done

yukezhu commented on robosuite/controllers/config/default_baxter.json:
```diff
@@ -0,0 +1,10 @@
+  "output_min": -1,
+  "kv": [8.0, 7.0, 6.0, 4.0, 2.0, 0.5, 0.1],
+  "velocity_limits": [-1,1],
+  "interpolation": null
```
add comma to the end of this line (same for the other jsons)

cremebrule replied:
json standard doesn't allow for trailing commas at the end of the dict. Tried and it brought up a parsing error

cremebrule commented on README.md:
```diff
@@ -60,7 +60,7 @@ First download MuJoCo 2.0 ([Linux](https://www.roboti.us/download/mujoco200_linu
    This will also install our library as an editable package, such that local changes will be reflected elsewhere without having to reinstall the package.
 
-3. (Optional) We also provide add-on functionalities, such as [OpenAI Gym](https://github.com/openai/gym) [interfaces](robosuite/wrappers/gym_wrapper.py), [inverse kinematics controllers](robosuite/wrappers/ik_wrapper.py) powered by [PyBullet](http://bulletphysics.org), and [teleoperation](robosuite/scripts/demo_spacemouse_ik_control.py) with [SpaceMouse](https://www.3dconnexion.com/products/spac ...
+3. (Optional) We also provide add-on functionalities, such as [OpenAI Gym](https://github.com/openai/gym) [interfaces](robosuite/wrappers/gym_wrapper.py), [multiple controllers](robosuite/controllers/README.md), including an [inverse kinematics controller](robosuite/controllers/ee_ik.py) powered by [PyBullet](http://bulletphysics.org), and [teleoperation](robosuite/scripts/demo_device_control.py) ...
```
done. Mentioned OSC in the main features enumeration section at the top, since it's not an "optional" add on.

cremebrule commented on robosuite/controllers/README.md:
```diff
@@ -0,0 +1,43 @@
+|    End Effector Inverse Kinematics (EE_IK)  		|    7 (x, y, z, i, j, k, w) 		|  
+|    Joint Position (JOINT_IMP)    					|    n (robot joint dimension)      |  
+|    Joint Velocity (JOINT_VEL)    					|    n (robot joint dimension)      |  
+|    Joint Torque (JOINT_TORQUE)   					|    n (robot joint dimension)      |  
```
done

yukezhu commented:
TODO: add additional dependencies (e.g., scipy) to requirements.txt

yukezhu commented:
> TODO: add additional dependencies (e.g., scipy) to requirements.txt

Looks like scipy is only used for `scipy.linalg.inv` in the interpolators. Can we change it to use `numpy.linalg.inv` instead?

roberto-martinmartin commented on robosuite/controllers/controller_factory.py:
```diff
@@ -0,0 +1,61 @@
+    Creates a Controller instance with the provided name and relevant params.
+
+    Args:
+        name: the name of the controller. Must be one of: {JOINT_IMP, JOINT_TORQUE, JOINT_VEL, EE_POS, EE_POS_ORI, EE_IK}
```
JOINT_TOR?

yukezhu replied:
fixed

yukezhu commented:
@amandlek @cremebrule Currently, `pybullet` (a heavy dependency) is required by default, which is only used for `controllers/ee_ik.py` and get called in `controllers/__init__.py`. Can we make this IK controller optional to keep the essential library requirements minimal?

cremebrule commented:
accidentally submitted wrong commit message for most recent. It should be:

-Fixed naming for other instances of JOINT_TORQUE -> JOINT_TOR
-Removed replaced scipy.linalg.inv -> np.linalg.inv
-Updated mujoco requirement to 2.0.2.9

yukezhu commented on robosuite/scripts/demo_device_ik_control.py:
```diff
@@ -89,7 +101,9 @@
         # rotate the gripper so we can see it easily
         if env.mujoco_robot.name == 'sawyer':
-            env.set_robot_joint_positions([0, -1.18, 0.00, 2.18, 0.00, 0.57, 1.5708])
+            # TODO: Confirm that this is no longer necessary
```
can we follow up on this TODO?

cremebrule replied:
this file is no longer relevant; it's been refactored as `demo_device_control.py` since it also includes osc demo functionality

yukezhu replied:
Is the plan to remove this demo script and replace it with a new demo script for all controllers?

cremebrule replied:
Yes; this PR doesn't have `demo_device_ik_control.py` in its repo -- it only has `demo_device_control.py`

cremebrule replied:
I don't think it makes sense to have a demo device script for all controllers since I'm not sure how / why users would want to try to directly control joint torques / velocities / positions? Unless you had specific thoughts in mind

yukezhu replied:
Great, `demo_device_control.py` is what exactly I want for replacement of `demo_device_ik_control.py`

amandlek commented:
@cremebrule @roberto-martinmartin I noticed that there are `position_limits` and orientation_limits` arguments for the OSC controllers, but they are unused by default. Do you know what the purpose of these is? I'm a little confused, since I thought restricting and scaling the action space of [-1, 1] would already restrict the motion of the arm.

cremebrule commented:
> @cremebrule @roberto-martinmartin I noticed that there are `position_limits` and `orientation_limits` arguments for the OSC controllers, but they are unused by default. Do you know what the purpose of these is? I'm a little confused, since I thought restricting and scaling the action space of [-1, 1] would already restrict the motion of the arm.

Since OSC is normally used with delta values as inputs, there's no explicit limiting of the actual position being requested. So at least for the position limits, it's the absolute range (in world coordinates) that's allowed. I believe orientation limits do the same thing, but @roberto-martinmartin should verify I'm not misinterpreting the code

roberto-martinmartin commented:
it looks good to me too

amandlek commented:
> > @cremebrule @roberto-martinmartin I noticed that there are `position_limits` and `orientation_limits` arguments for the OSC controllers, but they are unused by default. Do you know what the purpose of these is? I'm a little confused, since I thought restricting and scaling the action space of [-1, 1] would already restrict the motion of the arm.
> 
> Since OSC is normally used with delta values as inputs, there's no explicit limiting of the actual position being requested. So at least for the position limits, it's the absolute range (in world coordinates) that's allowed. I believe orientation limits do the same thing, but @roberto-martinmartin should verify I'm not misinterpreting the code

@roberto-martinmartin could you please confirm?

amandlek commented:
@cremebrule @yukezhu I think we should update the README for the repo before we merge, to emphasize new support for these different control interfaces. We should also tag the previous master branch so that when we complete this merge, people will know we've changed some conventions (for example, we are eliminating backwards compatibility with old joint velocity controller and RoboTurk-v1 dataset). 

For another README update - I think it would be good to have some picture of the environments with the Franka Arm (and some visual depiction of different controller interfaces). If you would rather wait for an official v2 release with corresponding white paper, we could push this to later though.

cremebrule commented:
> @cremebrule @yukezhu I think we should update the README for the repo before we merge, to emphasize new support for these different control interfaces. We should also tag the previous master branch so that when we complete this merge, people will know we've changed some conventions (for example, we are eliminating backwards compatibility with old joint velocity controller and RoboTurk-v1 dataset).
> 
> For another README update - I think it would be good to have some picture of the environments with the Franka Arm (and some visual depiction of different controller interfaces). If you would rather wait for an official v2 release with corresponding white paper, we could push this to later though.

Definitely agree with the former point. In terms of the latter, I think it's best to wait to overhaul the README until v2 (with the arm refactoring) is released, since a lot of things are going to need to be updated / explained in READMEs at that point. Also, no point in uploading photos with just the single Franka arm right now when there's gonna be a bunch of additional new (multi-arm) environments to show off as well with the immediate proceeding PR ;)

yukezhu commented:
I agreed with the former. For the latter, the reason I didn't update the
gallery image is that there are two things we'd like to do before
generating the new screenshots:
1. Figure out the support of textures of the robot arms, e.g., the textured
Franka arm as in Sapien <https://sapien.ucsd.edu/publication>;
2. Try with a new renderer of better quality. It seems that the Unity
plugin wasn't updated with MuJoCo 2.0 yet. We might have to look for new
options.

On Sat, Mar 28, 2020 at 9:01 PM cremebrule <notifications@github.com> wrote:

> @cremebrule <https://github.com/cremebrule> @yukezhu
> <https://github.com/yukezhu> I think we should update the README for the
> repo before we merge, to emphasize new support for these different control
> interfaces. We should also tag the previous master branch so that when we
> complete this merge, people will know we've changed some conventions (for
> example, we are eliminating backwards compatibility with old joint velocity
> controller and RoboTurk-v1 dataset).
>
> For another README update - I think it would be good to have some picture
> of the environments with the Franka Arm (and some visual depiction of
> different controller interfaces). If you would rather wait for an official
> v2 release with corresponding white paper, we could push this to later
> though.
>
> Definitely agree with the former point. In terms of the latter, I think
> it's best to wait to overhaul the README until v2 (with the arm
> refactoring) is released, since a lot of things are going to need to be
> updated / explained in READMEs at that point. Also, no point in uploading
> photos with just the single Franka arm right now when there's gonna be a
> bunch of additional new (multi-arm) environments to show off as well with
> the immediate proceeding PR ;)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/StanfordVL/robosuite/pull/63#issuecomment-605555209>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJDEAL52335HRFQRXIS4U3RJ3B2RANCNFSM4K26ZEHQ>
> .
>

roberto-martinmartin commented on robosuite/environments/sawyer_stack.py:
```diff
@@ -183,6 +203,10 @@ def _load_model(self):
 
+        # reset initial joint positions (gets reset in sim during super() call in _reset_internal)
+        self.init_qpos = np.array([-0.5538, -0.8208, 0.4155, 1.8409, -0.4955, 0.6482, 1.9628])
+        self.init_qpos += np.random.randn(self.init_qpos.shape[0]) * 0.02
```
@yukezhu  @amandlek , this is to initialize the arm in a slightly different position every episode. I found it useful to increase robustness, specially for sim2real (it is some kind of domain randomization in the initial joint configuration assuming the robot won't start exactly where it starts in sim). Should we make it cleaner? Should we remove it?

cremebrule replied:
I think it's a nice feature to have. With the new arm refactoring it can easily be passed as an (optional) argument at the top level

roberto-martinmartin commented:
> I agreed with the former. For the latter, the reason I didn't update the gallery image is that there are two things we'd like to do before generating the new screenshots: 1. Figure out the support of textures of the robot arms, e.g., the textured Franka arm as in Sapien <https://sapien.ucsd.edu/publication>; 2. Try with a new renderer of better quality. It seems that the Unity plugin wasn't updated with MuJoCo 2.0 yet. We might have to look for new options.
> […](#)
> On Sat, Mar 28, 2020 at 9:01 PM cremebrule ***@***.***> wrote: @cremebrule <https://github.com/cremebrule> @yukezhu <https://github.com/yukezhu> I think we should update the README for the repo before we merge, to emphasize new support for these different control interfaces. We should also tag the previous master branch so that when we complete this merge, people will know we've changed some conventions (for example, we are eliminating backwards compatibility with old joint velocity controller and RoboTurk-v1 dataset). For another README update - I think it would be good to have some picture of the environments with the Franka Arm (and some visual depiction of different controller interfaces). If you would rather wait for an official v2 release with corresponding white paper, we could push this to later though. Definitely agree with the former point. In terms of the latter, I think it's best to wait to overhaul the README until v2 (with the arm refactoring) is released, since a lot of things are going to need to be updated / explained in READMEs at that point. Also, no point in uploading photos with just the single Franka arm right now when there's gonna be a bunch of additional new (multi-arm) environments to show off as well with the immediate proceeding PR ;) — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#63 (comment)](https://github.com/StanfordVL/robosuite/pull/63#issuecomment-605555209)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJDEAL52335HRFQRXIS4U3RJ3B2RANCNFSM4K26ZEHQ> .

If we plan to do a push and merge all new functionalities (new robots, new controllers, new tasks, new multirobot support) soon, then, we should update the readme all at once. In that case, we should be fast in creating the whitepaper in the next month because it will be confusing to have the old robosuite publication linked to all these new functionalities that were not part of that. 

Wrt the textured Franka, I will take a look today and update the meshes if I find them.

Wrt new renderer, that can take longer. I would not block on that.
